File: flutter_client/lib/models/agent_event.dart
// ABOUTME: Domain events emitted by the agent for client consumption.
// ABOUTME: Abstracts away Agent SDK implementation details.

import 'dart:convert';
import 'geo_feature.dart';

/// Base class for all client events from the worker
abstract class AgentEvent {
  const AgentEvent();

  /// Parse a client event from JSON
  factory AgentEvent.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String?;

    switch (type) {
      case 'assistant_thinking':
        return const AssistantThinkingEvent();
      case 'assistant_message_chunk':
        return AssistantMessageChunkEvent(
          chunk: json['chunk'] as String? ?? '',
        );
      case 'assistant_message_complete':
        return const AssistantMessageCompleteEvent();
      case 'tool_start':
        return ToolStartEvent(
          toolId: json['toolId'] as String,
          toolName: json['toolName'] as String,
          input: json['input'] as Map<String, dynamic>? ?? {},
        );
      case 'tool_result':
        return ToolResultEvent(
          toolId: json['toolId'] as String,
          toolName: json['toolName'] as String,
          result: json['result'] as Map<String, dynamic>?,
          error: json['error'] as String?,
        );
      case 'subagent_started':
        return SubagentStartedEvent(
          toolId: json['toolId'] as String,
          agentName: json['agentName'] as String,
        );
      case 'subagent_completed':
        return SubagentCompletedEvent(
          toolId: json['toolId'] as String,
          agentName: json['agentName'] as String,
          error: json['error'] as String?,
        );
      case 'geo_feature':
        return GeoFeatureEvent(
          feature: GeoFeature.fromJson(json),
        );
      case 'feature_removed':
        return FeatureRemovedEvent(
          featureId: json['featureId'] as String,
          label: json['label'] as String? ?? 'Unknown',
        );
      case 'error':
        return ErrorEvent(
          errorType: json['errorType'] as String? ?? 'unknown',
          message: json['message'] as String? ?? 'Unknown error',
        );
      case 'done':
        return const DoneEvent();
      default:
        return UnknownEvent(type: type ?? 'unknown', data: json);
    }
  }

  /// Parse from SSE data string
  factory AgentEvent.fromSSEData(String data) {
    final json = jsonDecode(data) as Map<String, dynamic>;
    return AgentEvent.fromJson(json);
  }
}

/// Assistant is thinking (before first content arrives)
class AssistantThinkingEvent extends AgentEvent {
  const AssistantThinkingEvent();
}

/// Assistant message chunk (incremental text)
class AssistantMessageChunkEvent extends AgentEvent {
  final String chunk;

  const AssistantMessageChunkEvent({required this.chunk});
}

/// Assistant message is complete
class AssistantMessageCompleteEvent extends AgentEvent {
  const AssistantMessageCompleteEvent();
}

/// Tool execution started
class ToolStartEvent extends AgentEvent {
  final String toolId;
  final String toolName;
  final Map<String, dynamic> input;

  const ToolStartEvent({
    required this.toolId,
    required this.toolName,
    required this.input,
  });
}

/// Tool execution completed with result or error
class ToolResultEvent extends AgentEvent {
  final String toolId;
  final String toolName;
  final Map<String, dynamic>? result;
  final String? error;

  const ToolResultEvent({
    required this.toolId,
    required this.toolName,
    this.result,
    this.error,
  });

  bool get isError => error != null;
}

/// Sub-agent started
class SubagentStartedEvent extends AgentEvent {
  final String toolId;
  final String agentName;

  const SubagentStartedEvent({
    required this.toolId,
    required this.agentName,
  });
}

/// Sub-agent completed
class SubagentCompletedEvent extends AgentEvent {
  final String toolId;
  final String agentName;
  final String? error;

  const SubagentCompletedEvent({
    required this.toolId,
    required this.agentName,
    this.error,
  });

  bool get isError => error != null;
}

/// Geographic feature extracted from tool result
class GeoFeatureEvent extends AgentEvent {
  final GeoFeature feature;

  const GeoFeatureEvent({required this.feature});
}

/// Geographic feature removed from map
class FeatureRemovedEvent extends AgentEvent {
  final String featureId;
  final String label;

  const FeatureRemovedEvent({
    required this.featureId,
    required this.label,
  });
}

/// Error occurred during processing
class ErrorEvent extends AgentEvent {
  final String errorType; // 'tool_error', 'api_error', 'stream_error'
  final String message;

  const ErrorEvent({
    required this.errorType,
    required this.message,
  });
}

/// Stream is complete
class DoneEvent extends AgentEvent {
  const DoneEvent();
}

/// Unknown event type (for forward compatibility)
class UnknownEvent extends AgentEvent {
  final String type;
  final Map<String, dynamic> data;

  const UnknownEvent({required this.type, required this.data});
}

File: react_client/src/models/agent_event.ts
// ABOUTME: Domain events emitted by the agent for client consumption.
// ABOUTME: Uses discriminated unions for type-safe event handling.

import { type GeoFeature, parseGeoFeature } from "./geo_feature";

export type AgentEvent =
  | { type: "assistant_thinking" }
  | { type: "assistant_message_chunk"; chunk: string }
  | { type: "assistant_message_complete" }
  | {
      type: "tool_start";
      toolId: string;
      toolName: string;
      input: Record<string, unknown>;
    }
  | {
      type: "tool_result";
      toolId: string;
      toolName: string;
      result?: Record<string, unknown>;
      error?: string;
    }
  | { type: "subagent_started"; toolId: string; agentName: string }
  | {
      type: "subagent_text_chunk";
      toolId: string;
      agentName: string;
      chunk: string;
    }
  | {
      type: "subagent_completed";
      toolId: string;
      agentName: string;
      error?: string;
    }
  | { type: "geo_feature"; feature: GeoFeature }
  | { type: "feature_removed"; featureId: string; label: string }
  | { type: "error"; errorType: string; message: string }
  | { type: "done" }
  | { type: "unknown"; eventType: string; data: Record<string, unknown> };

export function parseAgentEvent(json: Record<string, unknown>): AgentEvent {
  const eventType = json.type as string | undefined;

  switch (eventType) {
    case "assistant_thinking":
      return { type: "assistant_thinking" };

    case "assistant_message_chunk":
      return {
        type: "assistant_message_chunk",
        chunk: (json.chunk as string) ?? "",
      };

    case "assistant_message_complete":
      return { type: "assistant_message_complete" };

    case "tool_start":
      return {
        type: "tool_start",
        toolId: json.toolId as string,
        toolName: json.toolName as string,
        input: (json.input as Record<string, unknown>) ?? {},
      };

    case "tool_result":
      return {
        type: "tool_result",
        toolId: json.toolId as string,
        toolName: json.toolName as string,
        result: json.result as Record<string, unknown> | undefined,
        error: json.error as string | undefined,
      };

    case "subagent_started":
      return {
        type: "subagent_started",
        toolId: json.toolId as string,
        agentName: json.agentName as string,
      };

    case "subagent_text_chunk":
      return {
        type: "subagent_text_chunk",
        toolId: json.toolId as string,
        agentName: json.agentName as string,
        chunk: (json.chunk as string) ?? "",
      };

    case "subagent_completed":
      return {
        type: "subagent_completed",
        toolId: json.toolId as string,
        agentName: json.agentName as string,
        error: json.error as string | undefined,
      };

    case "geo_feature":
      return {
        type: "geo_feature",
        feature: parseGeoFeature(json),
      };

    case "feature_removed":
      return {
        type: "feature_removed",
        featureId: json.featureId as string,
        label: (json.label as string) ?? "Unknown",
      };

    case "error":
      return {
        type: "error",
        errorType: (json.errorType as string) ?? "unknown",
        message: (json.message as string) ?? "Unknown error",
      };

    case "done":
      return { type: "done" };

    default:
      return {
        type: "unknown",
        eventType: eventType ?? "unknown",
        data: json,
      };
  }
}

export function parseAgentEventFromSSE(data: string): AgentEvent {
  const json = JSON.parse(data) as Record<string, unknown>;
  return parseAgentEvent(json);
}
